### ISSUE TITLE
Domain.wait/notify for Multi-domain Coordination

### ISSUE THREAD

**Issue by dlreeves** [ocamllabs/ocaml-multicore#117]
I work on the Hack type checker which is written in OCaml. I'm doing some initial investigation into multicore OCaml to see if it would be a suitable replacement down the line for our current method of parallelization via spawning multiple processes. We have a master process that farms out work to several worker processes. The worker processes essentially wait for jobs to execute by using unix select.

When translating this architecture to the multi-domain world one of the first things I encountered was trying to determine how to have a worker efficiently wait for jobs to execute. Ideally I would like to write something like:

```
while true do
  if job_queue_is_empty then
    Domain.wait ()
  else
    (* do work *)
done
```

Then the master could let a worker know there are jobs ready to execute by calling `Domain.notify domain_id`.

Is there already a way of achieving this level of coordination between domains? If not are there any plans to add such primitives?

**Comment by kayceesrk** [comment]
Hi @dlreeves. `Domain.wait` and `Domain.notify` seem like an abstraction that we should provide in multicore. For now, a polling mechanism such as [0] should do the trick. I have added a separate issue to track this feature #118. 

For the job queues, I'd suggest https://github.com/kayceesrk/lockfree which provides a lock-free queue implementation. You can install this library using [multicore-opam remote](https://github.com/ocamllabs/multicore-opam).

[0] https://github.com/kayceesrk/kcas/blob/master/src/kcas_backoff.ml

**Comment by stedolan** [comment]
We're definitely committed to providing a bunch of synchronisation primitives, including both low-level domain-based ones (mainly useful for projects like Hack that already have a threading model and just want to port it to multicore), and higher-level task-parallel ones.

Having said that, it probably won't be exactly `Domain.wait` and `Domain.notify`. With `wait`/`notify`-style interfaces, it's very difficult to get rid of the race condition between `job_queue_is_empty` and `Domain.wait ()` (what if the notification arrives just before the wait begins?). But there will definitely be a way of implementing this pattern.

**Comment by kayceesrk** [comment]
@dlreeves, @dhil is interested in looking at parallelizing Hack using multicore. I remember you mentioning that you had clear info about where changes were needed. If you have some notes, could you share them?

**Comment by kayceesrk** [comment]
Fixed by #144. See [domain.mli](https://github.com/ocamllabs/ocaml-multicore/blob/master/stdlib/domain.mli#L29).
